presto-main/src/main/java/io/prestosql/operator/MultiChannelGroupByHash.java

```diff
@@ -272,6 +277,74 @@ public int getCapacity()
         return hashCapacity;
     }
 
+    private int putIfAbsent(int firstPosition, int positions, Page page, @Nullable BlockBuilder groupIds)
+    {
+        /*
+         * Instead of looking for a match on a per-position basis, we execute steps in a "vectorized" way,
+         * so that all cpu mechanisms (caching, out-of-order execution, etc.) are fully utilized.
+         */
+        long[] rawHashes = new long[positions];
+        int[] hashPositions = new int[positions];
+        long[] groupAddresses = new long[positions];
+
+        for (int i = 0; i < positions; i++) {
+            rawHashes[i] = hashGenerator.hashPosition(firstPosition + i, page);
+        }
+
+        for (int i = 0; i < positions; i++) {
+            hashPositions[i] = (int) getHashPosition(rawHashes[i], mask);
+        }
+
+        for (int i = 0; i < positions; i++) {
+            groupAddresses[i] = groupAddressByHash[hashPositions[i]];
+        }
+
+        for (int i = 0; i < positions; ++i) {
+            long rawHash = rawHashes[i];
+            int hashPosition = hashPositions[i];
+            long groupAddress = groupAddresses[i];
+
+            // look for an empty slot or a slot containing this key
+            int groupId = -1;
+            while (groupAddress != -1 || groupAddressByHash[hashPosition] != -1) {
```

Comment:
Double check me on this, but it looks like groupAddress always equals groupAddressByHash[hashPosition]. If so, that would make this second check and the below if-condition redundant?

Reply:
> Double check me on this, but it looks like groupAddress always equals groupAddressByHash[hashPosition]

No. When group is inserted into hash map, then precomputed `groupAddresses` might not be valid then they are `-1`.

Precomputing `groupAddresses` helps mostly when there are preexiting collisions or if there is a matching group.

Reply:
Are you sure? Please double check here. I stared at this for a bit, and I was pretty sure I couldn't see how those values would diverge in this loop.

Reply:
see `io.trino.operator.MultiChannelGroupByHash#addNewGroup`, it changes `groupAddressByHash`